Feature: Checkpointing for T8codeMesh

jmark opened the pull request:
This PR adds checkpointing for `T8codeMesh`. By this, routines like `save_mesh` and `load_mesh` are supported.

Closes #2044

benegee commented on src/meshes/mesh_io.jl:
```diff
@@ -225,11 +225,88 @@ function save_mesh_file(mesh::P4estMesh, output_directory, timestep, mpi_paralle
-    error("Mesh file output not supported yet for `T8codeMesh`.")
+# This routine works for both, serial and MPI parallel mode. The forest
+# information is collected on all ranks and then gathered by the root rank.
+# Since only the `levels` array of unsigned bytes is basicially independent per
+# rank it is not worth the effort to have a collective write to the HDF5 file.
```
Out of curiosity: This mean that all other data is present and synchronized on all ranks?

jmark replied:
No, the meaning behind this comment is, that only the `levels` array together with the `num_elements_per_tree` array is sufficient in order to reconstruct the whole forest. Since both arrays are not extremely data intensive, I decided against a collective write from all processes and instead simply gather everything on the root rank.

jmark replied:
Made the comment clearer.

benegee commented on src/meshes/t8code_mesh.jl:
```diff
@@ -108,6 +111,153 @@ Main mesh constructor for the `T8codeMesh` wrapping around a given t8code
+            vertices[24] = tree_node_coordinates[3, N, N, N, i]
+        end
+
+        t8_cmesh_set_tree_vertices(cmesh, i - 1, vertices, 3)
```
`3` here means three vertices (consisting of 3 doubles each), right?
But shouldn't there be 4 or 8, respectively?

I had difficulties with this already in the past...

jmark replied:
Good catch! Fixed.

benegee commented on src/meshes/t8code_mesh.jl:
```diff
@@ -108,6 +111,153 @@ Main mesh constructor for the `T8codeMesh` wrapping around a given t8code
+    if mpi_isparallel()
+        forest = partition(forest)
+    end
+
```
How about a final check that the number of elements reached matches with the given `nelements`?

jmark replied:
Good idea! Added.

benegee commented on src/meshes/t8code_mesh.jl:
```diff
@@ -810,32 +972,32 @@ function count_interfaces(mesh::T8codeMesh)
 
-    max_level = t8_forest_get_maxlevel(mesh.forest) #UInt64
-    max_tree_num_elements = UInt64(2^ndims(mesh))^max_level
+    max_level = t8_forest_get_maxlevel(forest) #UInt64
```
Should the comment indicate that `max_level` is an unsigned?

jmark replied:
Probably. Do not remember anymore, why I put that comment there. Removed it.

benegee commented on src/meshes/t8code_mesh.jl:
```diff
@@ -982,12 +1144,14 @@ function fill_mesh_info!(mesh::T8codeMesh, interfaces, mortars, boundaries,
     # Helper variables to compute unique global MPI interface/mortar ids.
     max_level = t8_forest_get_maxlevel(mesh.forest) #UInt64
-    max_tree_num_elements = UInt64(2^ndims(mesh))^max_level
+    max_tree_num_elements = (UInt128(2)^ndims(mesh))^max_level
```
Was there a specific reason for converting `2` instead of coverting `2^ndims` as done above?

jmark replied:
None in particular. I think the result is the same. Maybe in the future we need support for thousands of dimensions ...

benegee commented on src/meshes/t8code_mesh.jl:
```diff
@@ -810,32 +972,32 @@ function count_interfaces(mesh::T8codeMesh)
-    max_level = t8_forest_get_maxlevel(mesh.forest) #UInt64
-    max_tree_num_elements = UInt64(2^ndims(mesh))^max_level
+    max_level = t8_forest_get_maxlevel(forest) #UInt64
+    max_tree_num_elements = UInt128(2^ndims)^max_level
```
:ok_hand:

benegee commented on src/meshes/t8code_mesh.jl:
```diff
@@ -1263,6 +1428,64 @@ function fill_mesh_info!(mesh::T8codeMesh, interfaces, mortars, boundaries,
+end
+
+function get_cmesh_info(cmesh::Ptr{t8_cmesh}, ndims)
+    num_trees = t8_cmesh_get_num_trees(cmesh)
```
This and all other data are global?

jmark replied:
Yes! Every MPI ranks has a full copy of `cmesh`. Added a comment explaining that.

benegee reviewed (changes requested):
Great feature. Thank a lot!

jmark commented on examples/t8code_2d_dgsem/elixir_advection_amr_unstructured_flag.jl:
```diff
@@ -54,19 +54,29 @@ analysis_callback = AnalysisCallback(semi, interval = analysis_interval,
                            adapt_initial_condition = true,
-                           adapt_initial_condition_only_refine = true)
+                           adapt_initial_condition_only_refine = true,
+                           dynamic_load_balancing = false)
```
Explained it in a comment.

jmark commented on src/meshes/t8code_mesh.jl:
```diff
@@ -108,6 +111,153 @@ Main mesh constructor for the `T8codeMesh` wrapping around a given t8code
+
+    # Use linear geometry for now. There is no real Lagrange geometry
+    # implementation yet in t8code.
+    Trixi.linear_geom = Trixi.t8_geometry_linear_new(2)
```
Goodl catch! Fixed.

JoshuaLampert commented on src/meshes/mesh_io.jl:
```diff
@@ -225,11 +225,88 @@ function save_mesh_file(mesh::P4estMesh, output_directory, timestep, mpi_paralle
+# rank it is not worth the effort to have a collective write to the HDF5 file.
+# Instead, `levels` gets gathered by the root rank and written in serial.
+function save_mesh_file(mesh::T8codeMesh, output_directory, timestep,
+                        mpi_parallel::Any)
```
```suggestion
                        mpi_parallel)
```
Or maybe also `Union{False, True}`?

jmark replied:
Good point! I implemented Benedicts suggestion.

JoshuaLampert commented on src/meshes/t8code_mesh.jl:
```diff
@@ -108,6 +111,153 @@ Main mesh constructor for the `T8codeMesh` wrapping around a given t8code
+    end
+
+    # Connect the coarse mesh elements.
+    for i in 1:length(elemIDs)
```
```suggestion
    for i in eachindex(elemIDs)
```
We usually prefer `eachindex` over `1:length`.

JoshuaLampert reviewed (commented):
Nice! Thanks a lot!

jmark commented on src/meshes/t8code_mesh.jl:
```diff
@@ -108,6 +111,153 @@ Main mesh constructor for the `T8codeMesh` wrapping around a given t8code
 `forest` object. This constructor is typically called by other `T8codeMesh`
 constructors.
 
+# Arguments
+- `forest`: Pointer to a t8code forest.
+- `boundary_names`: List of boundary names.
+- `polydeg::Integer`: Polynomial degree used to store the geometry of the mesh.
+                      The mapping will be approximated by an interpolation polynomial
+                      of the specified degree for each tree.
+- `mapping`: A function of `NDIMS` variables to describe the mapping that transforms
+             the imported mesh to the physical domain. Use `nothing` for the identity map.
+"""
```
Fixed!

JoshuaLampert commented on examples/t8code_2d_dgsem/elixir_advection_restart.jl:
```diff
@@ -0,0 +1,43 @@
+# create a restart file
+
+elixir_file = "elixir_advection_extended.jl"
+restart_file = "restart_000021.h5"
```
```suggestion
restart_file = "restart_000000021.h5"
```
Due to the breaking change in v0.8.

JoshuaLampert commented on examples/t8code_2d_dgsem/elixir_advection_restart_amr.jl:
```diff
@@ -0,0 +1,62 @@
+# create a restart file
+
+elixir_file = "elixir_advection_extended.jl"
+restart_file = "restart_000021.h5"
```
```suggestion
restart_file = "restart_000000021.h5"
```

JoshuaLampert commented on examples/t8code_3d_dgsem/elixir_advection_restart.jl:
```diff
@@ -0,0 +1,41 @@
+# Note: If you get a restart file from somewhere else, you need to provide
+# appropriate setups in the elixir loading a restart file
+
+restart_filename = joinpath("out", "restart_000010.h5")
```
```suggestion
restart_filename = joinpath("out", "restart_000000010.h5")
```

JoshuaLampert commented on src/meshes/mesh_io.jl:
```diff
@@ -225,11 +225,89 @@ function save_mesh_file(mesh::P4estMesh, output_directory, timestep, mpi_paralle
+
+    # Determine file name based on existence of meaningful time step.
+    if timestep > 0
+        filename = joinpath(output_directory, @sprintf("mesh_%06d.h5", timestep))
```
```suggestion
        filename = joinpath(output_directory, @sprintf("mesh_%09d.h5", timestep))
```

JoshuaLampert commented on examples/t8code_3d_dgsem/elixir_advection_restart.jl:
```diff
@@ -0,0 +1,41 @@
+
```
```suggestion
```

JoshuaLampert commented on src/meshes/t8code_mesh.jl:
```diff
@@ -101,6 +104,166 @@ function Base.show(io::IO, ::MIME"text/plain", mesh::T8codeMesh)
+"""
+    T8codeMesh(ndims, ntrees, nelements, tree_node_coordinates, nodes,
+               boundary_names, treeIDs, neighIDs, faces, duals,
+               orientations, levels, num_elements_per_tree))
```
```suggestion
               orientations, levels, num_elements_per_tree)
```

JoshuaLampert commented on src/meshes/t8code_mesh.jl:
```diff
@@ -810,32 +985,32 @@ function count_interfaces(mesh::T8codeMesh)
     local_num_mpi_mortars = 0
 
-    visited_global_mortar_ids = Set{UInt64}([])
+    visited_global_mortar_ids = Set{UInt128}([])
```
Can you explain why it is necessary to use `UInt128` instead of `UInt64` as before?

jmark replied:
Sure! The associated computation like [this](https://github.com/trixi-framework/Trixi.jl/blob/611a5761efbdce07a3c20cbc444e603a511c4f2e/src/meshes/t8code_mesh.jl#L1072)
can lead to very big numbers since
```
max_level = 29
max_tree_num_elements = UInt128(2^ndims)^max_level
```
is very big. With meshes of a small number of `cmesh` elements respectively trees (around 10 trees) the computation fits into a `UInt64`. In case of a bigger number of trees you get an integer overflow. This is actually a serious bug which I fix with this PR. The bug came not into effect before since I tested only for smaller meshes with a small number of trees. While developing this PR this bug bubbled up. As a matter of fact, the code did not blow. It led to a subtle mismatching of MPI interfaces in case of MPI parallel execution. With `UInt128` there is more than enough space for extremely large numbers of trees.

JoshuaLampert replied:
Great, thanks a lot!

JoshuaLampert commented on src/meshes/t8code_mesh.jl:
```diff
@@ -1092,6 +1255,21 @@ function fill_mesh_info!(mesh::T8codeMesh, interfaces, mortars, boundaries,
                     neighbor_level = t8_element_level(neighbor_scheme, neighbor_leaves[1])
 
+                    # Compute the `orientation` of the touching faces.
+                    if t8_element_is_root_boundary(eclass_scheme, element, iface) == 1
```
```suggestion
                    if t8_element_is_root_boundary(eclass_scheme, element, iface)
```
Doesn't `t8_element_is_root_boundary` return a boolean? We have
```julia
julia> true == 1
true
```
but this looks strange to me if it's a boolean already

jmark replied:
When I do this as you suggest, I always get the error message:
```
ERROR: TypeError: non-boolean (Int64) used in boolean context
```

JoshuaLampert replied:
Ah ok, so it does indeed return an integer and not a boolean, which is kind of strange in my opinion, but ok. Thanks!

JoshuaLampert commented on src/meshes/t8code_mesh.jl:
```diff
@@ -1263,6 +1441,66 @@ function fill_mesh_info!(mesh::T8codeMesh, interfaces, mortars, boundaries,
+end
+
+function get_cmesh_info(mesh::T8codeMesh)
+    @assert t8_forest_is_committed(mesh.forest) != 0
```
```suggestion
    @assert t8_forest_is_committed(mesh.forest)
```
Similar question to above.

jmark replied:
Same answer as above. :D

JoshuaLampert replied:
If the only possible values are 0 and 1 could be turn this into
```suggestion
    @assert t8_forest_is_committed(mesh.forest) == 1
```
for more consistency and I think a bit better readability?

jmark replied:
Fair enough.

JoshuaLampert reviewed (commented):
Do you think the MPI failures are really unrelated?

benegee commented:
> Do you think the MPI failures are really unrelated?

No, I am not sure. I just know that we had stalling CI jobs before. And, looking trough the recent failures, I think it was not always the same elixir.

JoshuaLampert commented:
>I get the feeling that the MPI tests are too big now and take too long. We probably have to split them up similar to the serial tests.

Yes, could be related to OOM issues, cf. https://github.com/trixi-framework/Trixi.jl/issues/1471.

jmark commented:
> > I get the feeling that the MPI tests are too big now and take too long. We probably have to split them up similar to the serial tests.
> 
> Yes, could be related to OOM issues, cf. #1471.

I could narrow it down. It has something to do with Julia 10.1.4. With Julia 10.1.2 it does not stall. Investigating ...

JoshuaLampert commented:
Are you able to reproduce the problem locally?

jmark commented:
Yes! With Julia 1.10.2 the t8code MPI tests run successfully. However, with Julia 1.10.4 the MPI test for `elixir_advection_restart.jl` 2D stalls for whatever reason. Running the elixir with MPI directly does not stall. No idea, right now, what's going on ...

sloede commented:
> Yes! With Julia 1.10.2 the t8code MPI tests run successfully. However, with Julia 1.10.4 the MPI test for `elixir_advection_restart.jl` 2D stalls for whatever reason. Running the elixir with MPI directly (not wrapped in a test set) does not stall. No idea, right now, what's going on ...

Are you sure it's related to the patch version bump? Are you using an identical Manifest.toml for both tests?

jmark commented:
> > Yes! With Julia 1.10.2 the t8code MPI tests run successfully. However, with Julia 1.10.4 the MPI test for `elixir_advection_restart.jl` 2D stalls for whatever reason. Running the elixir with MPI directly (not wrapped in a test set) does not stall. No idea, right now, what's going on ...
> 
> Are you sure it's related to the patch version bump? Are you using an identical Manifest.toml for both tests?

Yes! Working from the exact same project folder. Just pointing the Julia binary to either 1.10.2 or 1.10.4.

JoshuaLampert commented:
So it consistently stalls with Julia 1.10.4, but consistently works with Julia 1.10.2 in multiple runs? Did you monitor RAM usage during the simulation?

jmark commented:
> So it consistently stalls with Julia 1.10.4, but consistently works with Julia 1.10.2 in multiple runs? Did you monitor RAM usage during the simulation?

Yes! RAM usage is not out of ordinary.

jmark commented:
I think I found the bug causing the stalls in the MPI runs. It was a silent memory leak/segfault. I added the fixes in the last commit. Furthermore, I changed the t8code C interface a tiny bit to simplify the code on Trixi side. This PR has to wait for the next breaking t8code release and specifically for the merge of this PR: https://github.com/DLR-AMR/t8code/pull/1115.

I'll try to push for a major t8code release by the end of next week.

benegee commented:
t8code 3.0.0 has been released and @jmark already updated t8code_jll.jl. Does T8code.jl need an update? Or is this good to go?

benegee commented on src/auxiliary/t8code.jl:
```diff
@@ -14,30 +14,33 @@ function init_t8code()
-            # variable TRIXI_T8CODE_SC_FINALIZE exists.
-            @info "T8code.jl: `sc_finalize` will be called during shutdown of Trixi.jl."
-            MPI.add_finalize_hook!(T8code.Libt8.sc_finalize)
+        MPI.add_finalize_hook!() do
```
This lets downgrade test fail. MPI v0.20.0 is too old.

benegee replied:
Set to v.0.20.6 now.

jmark commented:
> t8code 3.0.0 has been released and @jmark already updated t8code_jll.jl. Does T8code.jl need an update as well?

Yes, indeed! Working on that.

benegee commented:
Concerning the failing invalidations check, see https://github.com/timholy/SnoopCompile.jl/issues/397

jmark commented:
@benegee As expected, several tests fail since some routines became obsolete with the breaking t8code 3.0.0 release. We have to iron out these issues first.

I suggest we fix these problems in this PR https://github.com/trixi-framework/Trixi.jl/pull/1939 first since the changes a minimal on Trixi side. Then merge in this PR into the checkpointing PR.

ranocha commented on src/auxiliary/t8code.jl:
```diff
@@ -14,30 +14,33 @@ function init_t8code()
-        # Initialize t8code with log level ERROR to prevent a lot of output in AMR simulations.
-        t8_init(T8code.Libt8.SC_LP_ERROR)
-
-        if haskey(ENV, "TRIXI_T8CODE_SC_FINALIZE")
```
Why did you remove the effect of this environment variable?

jmark replied:
I decided that `sc_finalize` should be called always. A warning to the user is issued if there are still unfreed t8code objects lying around when Trixi shuts down. This is a serious issue since t8code allocates several MPI related objects (shared memory arrays). When they are not properly freed before MPI shuts down the program crashes before exiting in an orderly fashion.

See the code a couple of lines below.

jmark replied:
Due to the way the GC works in Julia I cannot influence the order of shutting down. Or at least, I couldn't find an elegant and robust way how to achieve this.

ranocha replied:
Ok, thanks
